articles/compound-component-pattern-101.md

```diff
@@ -0,0 +1,148 @@
+  description: string
+  price: number
+}
+
+const Card: React.FC<CardProps> = ({
+  title,
+  photo,
+  description,
+  price,
+}) => {
+  return (
+    <div className="card">
+      <h2 className="card__title">{title}</h2>
+      <img className="card__photo" src={photo} alt="" />
+      <p className="card__description">{description}</p>
+      <span className="card__price">{price}</span>
+      <span className="card__add-to-cart">
+        {" "}
+        add to cart {"Imagine a card icon ..."}{" "}
+      </span>
+    </div>
+  );
+};
+
+export default Card
+```
+
+But then you look at the website's design and there is more than one type of card, you also have a card without a description : 
+
+![card-component-simple.jpeg](/images/articles/compound-component-pattern/card-component-simple-60.jpeg)
+
+Most of the time developers would add more complexity to their current component, making it more bloated than neccesary, like adding conditional rendering on the paragraph holding the description and also another conditional which will change the classes and apply a different css depending if the description argument was passed or not, or even WORSE there is the option of creating two distinct components for this two designs (one with the description, another one without it), this can easily be avoided if you create a compound component.
+
+Now, the compound approach will lead us to fragmenting our component into smaller components, one for each part of the card, this will make our code look like this:
+
+```ts 
+import React from "react";
+
+const CompoundCard = ({ children }: React.PropsWithChildren) => {
+  return <section className="card">{children}</section>;
+};
+
+const Header = ({ children }: React.PropsWithChildren) => {
+  return <h2 className="card__title">{children}</h2>;
+};
+
+const Photo: React.FC<{ photo: string; alt: string }> = ({ photo, alt }) => {
+  return <img className="card__photo" src={photo} alt={alt} />;
+};
+
+const Description = ({ children } : React.PropsWithChildren) => {
+  return <p className="card__description">{children}</p>;
+};
+
+const Price: React.FC<{ price: number }> = ({ price }) => {
+  return <span className="card__price">{price}</span>;
+};
+
+const AddToCart: React.FC = (second) => {
+  return (
+    <span className="card__add-to-cart">
+      add to cart {"Imagine a card icon ..."}
+    </span>
+  );
+};
+
+export default CompoundCard;
+```
+
+Now, we will take advantage of a Javascript feature : functions are also objects. we are going to attatch each of our component parts to the main "CompoundCard" component as an object property, just like this: 
+
+```ts 
+CompoundCard.Header = Header
+CompoundCard.Photo = Photo
+CompoundCard.Description = Description
+CompoundCard.Price = Price
+CompoundCard.AddToCart = AddToCart
+```
+
+It might look weird if you are not familiarized with how javascript works, but in this implementation we can start using the compound card component with a more interesting API 
```

Comment:
```suggestion
It might look weird if you are not familiarized with how javascript works, but in this implementation we can start using the compound card component with a more interesting API.
```

Reply:
When you mean API are you meaing the fact of threating the CompoundCard as an object? is thtat what you called API? Because that could likely be more a technique than a API but I am not sure, could you check this out?